compiler/qsc_qasm/src/semantic/lowerer.rs

```diff
@@ -1546,8 +1821,65 @@ impl Lowerer {
         // by all the QASM semantic analysis.
     }
 
+    fn index_into_qubit_register(op: semantic::GateOperand, index: u32) -> semantic::GateOperand {
+        let index = semantic::Index::Expr(semantic::Expr {
+            span: op.span,
+            kind: Box::new(semantic::ExprKind::Lit(semantic::LiteralKind::Int(
+                index.into(),
+            ))),
+            ty: Type::UInt(None, true),
+        });
+
+        match op.kind {
+            semantic::GateOperandKind::Expr(expr) => {
+                // Single qubits are allowed in a broadcast call.
+                if matches!(&expr.ty, Type::Qubit) {
+                    return semantic::GateOperand {
+                        span: op.span,
+                        kind: semantic::GateOperandKind::Expr(expr),
+                    };
+                }
+
+                assert!(matches!(expr.ty, Type::QubitArray(..)));
```

Comment:
This works but could just as easily be a match with `semantic::GateOperand` from two cases and a `_ => unreachable!(...)` for the rest.